File: 2025-05-MAY-GFG-POTD/May 03 - Prime list/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX = 100010;
bool isPrime[MAX];

// Sieve of Eratosthenes
void sieve() {
    fill(isPrime, isPrime + MAX, true);
    isPrime[0] = isPrime[1] = false;
    for (int i = 2; i * i < MAX; i++) {
        if (isPrime[i]) {
            for (int j = i * i; j < MAX; j += i) {
                isPrime[j] = false;
            }
        }
    }
}

struct Node {
    int val;
    Node* next;
    Node(int data) : val(data), next(nullptr) {}
};

int nearestPrime(int num) {
    if (isPrime[num]) return num;
    int left = num - 1;
    int right = num + 1;
    while (left >= 2 || right < MAX) {
        if (left >= 2 && isPrime[left]) return left;
        if (right < MAX && isPrime[right]) return right;
        left--;
        right++;
    }
    return 2;
}

Node* primeList(Node* head) {
    Node* curr = head;
    while (curr) {
        curr->val = nearestPrime(curr->val);
        curr = curr->next;
    }
    return head;
}

File: 2025-05-MAY-GFG-POTD/May 03 - Prime list/Solution.py
MAX = 100010
is_prime = [True] * MAX

# Sieve of Eratosthenes
def sieve():
    is_prime[0] = is_prime[1] = False
    for i in range(2, int(MAX**0.5) + 1):
        if is_prime[i]:
            for j in range(i*i, MAX, i):
                is_prime[j] = False

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def nearest_prime(num):
    if is_prime[num]:
        return num
    left = num - 1
    right = num + 1
    while left >= 2 or right < MAX:
        if left >= 2 and is_prime[left]:
            return left
        if right < MAX and is_prime[right]:
            return right
        left -= 1
        right += 1
    return 2

def prime_list(head):
    curr = head
    while curr:
        curr.val = nearest_prime(curr.val)
        curr = curr.next
    return head